devassistant.1
.\" Copyright Petr Hracek, 2013
.\"
.\" This page is distributed under GPL.
.\"
.TH DEVASSISTANT 1 2013-03-12 "" "Linux User's Manual"
.SH NAME
devassistant \- Developer assistant is a helper for all developers using (not-only) Fedora. It helps with creating and setting up basic projects in various languages, installing dependencies, setting up environment etc.
.SH SYNOPSIS
\fBdevassistant \fP c [-h] [-b [BUILD]] [-g [GITHUB]] [-e [ECLIPSE]] -n NAME [-v [VIM]]
.br
\fBdevassistant \fP cpp [-h] [-b [BUILD]] [-g [GITHUB]] [-e [ECLIPSE]] -n NAME [-v [VIM]]
.br
\fBdevassistant \fP java jsf [-h] [-g [GITHUB]] [-e [ECLIPSE]] -n NAME
.br
\fBdevassistant \fP java maven [-h] [-g [GITHUB]] [-e [ECLIPSE]] -n NAME
.br
\fBdevassistant \fP perl class [-h] [-g [GITHUB]] -n NAME [-v [VIM]]
.br
\fBdevassistant \fP perl dancer [-h] [--cgi] [-g [GITHUB]] -n NAME [--fastcgi]
.br
\fBdevassistant \fP php lamp [-h] [-g [GITHUB]] -n NAME [-r ROOTDB] [-v [VIM]]
.br
\fBdevassistant \fP python django [-h] [-g [GITHUB]] [-e [ECLIPSE]] -n NAME [-v [VIM]]
.br
\fBdevassistant \fP python flask [-h] [-g [GITHUB]] [-e [ECLIPSE]] -n NAME [-v [VIM]]
.br
\fBdevassistant \fP python lib [-h] [-g [GITHUB]] [-e [ECLIPSE]] -n NAME [-v [VIM]]
.br
\fBdevassistant \fP python pygtk [-h] [-g [GITHUB]] [-e [ECLIPSE]] -n NAME [-v [VIM]]
.br
\fBdevassistant \fP ruby rails [-h] [-g [GITHUB]] -n NAME [-v [VIM]]
.br
.SH GITHUB INTEGRATION
If github option is used for export projects to the GitHub repository then
user has to have already created account on the GitHub server.
Without that it is not possible to upload projects on the GitHub server

.SH DESCRIPTION
.B Developer assistant
is based on idea of per-{language/framework/...} "assistants" with hierarchical structure like this:

                  MainAssistant
                  /           \\
                 /             \\
              Python          Ruby
              /   \\            / \\
             /     \\          /   \\
          Django  Flask    Rails Sinatra

The basic task of each assistant is to create a project in location specified by
.B -n name
(either relative or absolute path) and install its dependencies via \fBYum\fP.
Optionally, if any environment setup is required,
.B devassistant
does that as well.
Most of the assistants have common arguments (see below), but there may also be some language specific options in some of the assistants.

.SH LANGUAGE SUPPORT
Currently supported languages are:
.IP
C/C++ language
.IP
Python language (Django, Flask , PyGTK (GTK3) and a custom Python library project)
.IP
Java language (JSF and a custom Maven project)
.IP
Perl language (Basic class and dancer framework)
.IP
PHP language (LAMP project)

.SH COMMON OPTIONS
.TP
.B \-h --help
print help for selected assistant
.TP
.B \-n --name
name of the project (relative or absolute path).
.TP
.B \-e --eclipse
import project into
.I eclipse 
environment. Uses ~/workspace workspace by default, which can be overriden by specifying another path.
.TP
.B \-g --github <username>
create a GitHub repo and push the initial sources there. Uses current username, which can be overriden by specifying a different GitHub username. The GitHub projectname is taken from
.B name
parameter..B \-b --build
generate a SPEC file, SRPM and mock-build RPM.
.TP
.B \-v --vim
install vim settings specific for this language (backs up your original vim settings)
.TP

.SH "SEE ALSO"
.BR devassistant-modify (1)


CHANGELOG
0.4.0
- Introduction of "preparer" assistants (use devassitant-prepare binary) that
  can checkout project from SCM and install it's dependencies according to
  .devassistant file
- Perl assistants, PHP assistant
- New way of dependency installation (first find dependencies via yum api,
  then show them to user and let him confirm, then install)
- Interactive commands now use zenity, if available, otherwise they fall
  back to terminal.
- New style of calling assistant/snippet sections
  (call: snippet.section; call: self.run_foo)
- Support for using SCLs easily (the "scl enable" command in yaml assistants)
- GitHub backend rewrite (can now create ssh keys and add them to the user,
  remembers authentication token)
- Some minor/less visible changes (logging improvement, bugfixing, cleanup)

0.3.1
- Abandoned plumbum, implemented own simpler and more suitable solution.
- Assistants are now sorted by name in usage output.
- Yum is executed via pkexec, so that users don't have to alter sudoers.

0.3.0
- Changelog starts here, although some versions were already released.
- At this point, devassistant can create C, C++, various Python, Java and
  Ruby projects
- Devassistant also has the ability to modify existing projects (import to
  eclipse so far)
- The features of created projects include (except for the source code itself)
-- Option to register project into eclipse
-- Option to push project to GitHub
-- Option to set up Vim
- Devassistant can install RPM dependencies
- GUI is not yet operational (although there is a first testing code)
- Yaml assistants have pretty nice set of features and you can implement pretty
  much anything in them (snippets can DRY the code, etc.)


devassistant.bash
# bash completion for devassistant

_devassistant_baseopts_c()
{
	echo ""
    local opts='--github --help --vim --build --eclipse --name'
    printf %s "$opts"
}

_devassistant_baseopts_python()
{
	echo ""
    local opts='--github --help --eclipse --name --vim'
    printf %s "$opts"
}

_devassistant_baseopts_java()
{
	echo ""
    local opts='--github --help --eclipse --name'
    printf %s "$opts"
}

_devassistant_baseopts_perl()
{
	echo ""
    local opts='--github --help --eclipse --name'
    printf %s "$opts"
}

_devassistant_baseopts_perl_dancer()
{
	echo ""
    local opts='--github --cgi --help --eclipse --name --fastcgi'
    printf %s "$opts"
}

_devassistant_baseopts_php()
{
	echo ""
    local opts='--github --help --eclipse --name --rootdb --vim'
    printf %s "$opts"
}

_devassistant()
{
    COMPREPLY=()
    local devassistant=$1 cur=$2 prev=$3 words=("${COMP_WORDS[@]}")
    declare -F _get_comp_words_by_ref &>/dev/null && \
        _get_comp_words_by_ref -n = cur prev words

    # Commands offered as completions
    local cmds=( c cpp java python perl php)

    local i c cmd subcmd
    for (( i=1; i < ${#words[@]}-1; i++ )) ; do
        [[ -n $cmd ]] && subcmd=${words[i]} && break
        # Recognize additional commands and aliases
        for c in ${cmds[@]} ; do
            [[ ${words[i]} == $c ]] && cmd=$c && break
        done
    done

	#echo "devassistant: $devassistant, cur: $cur, prev: $prev, word: $words, subcmd: $subcmd, cmd:$cmd"
    case $cmd in

        c)
			if [[ $prev == $cmd ]] ; then
				COMPREPLY=( $( compgen -W '--github --help --vim --build --eclipse --name' -- "$cur" ) )
				return 0
			fi
			if [[ $subcmd == -* ]] ; then
				COMPREPLY=( $( compgen -W '$( _devassistant_baseopts_c )' -- "$cur" ) )
				return 0
			fi
            ;;

        cpp)
			if [[ $prev == $cmd ]] ; then
				COMPREPLY=( $( compgen -W '--github --help --vim --build --eclipse --name' -- "$cur" ) )
				return 0
			fi
			if [[ $subcmd == -* ]] ; then
				COMPREPLY=( $( compgen -W '$( _devassistant_baseopts_c )' -- "$cur" ) )
				return 0
			fi
            ;;

        java)
            if [[ $prev == $cmd ]] ; then
                COMPREPLY=( $( compgen -W 'jsf maven' -- "$cur" ) )
                return 0
            fi
            case $subcmd in
                jsf)
					COMPREPLY=( $( compgen -W '$( _devassistant_baseopts_java )' -- "$cur" ) )
					return 0
					;;
                maven)
					COMPREPLY=( $( compgen -W '$( _devassistant_baseopts_java )' -- "$cur" ) )
					return 0
					;;
            esac
            return 0
            ;;

        perl)
            if [[ $prev == $cmd ]] ; then
                COMPREPLY=( $( compgen -W 'class dancer' -- "$cur" ) )
                return 0
            fi
            case $subcmd in
                class)
					COMPREPLY=( $( compgen -W '$( _devassistant_baseopts_perl )' -- "$cur" ) )
					return 0
					;;
                dancer)
					COMPREPLY=( $( compgen -W '$( _devassistant_baseopts_perl_dancer )' -- "$cur" ) )
					return 0
					;;
            esac
            return 0
            ;;


        python)
            if [[ $prev == $cmd ]] ; then
				COMPREPLY=( $( compgen -W 'lib django flask pygtk' -- "$cur" ) )
				return 0
			fi
            case $subcmd in
                lib)
					COMPREPLY=( $( compgen -W '$( _devassistant_baseopts_python )' -- "$cur" ) )
					return 0
					;;
                django)
					COMPREPLY=( $( compgen -W '$( _devassistant_baseopts_python )' -- "$cur" ) )
					return 0
					;;
                flask)
					COMPREPLY=( $( compgen -W '$( _devassistant_baseopts_python )' -- "$cur" ) )
					return 0
					;;
                pygtk)
					COMPREPLY=( $( compgen -W '$( _devassistant_baseopts_python )' -- "$cur" ) )
					return 0
					;;
            esac
            return 0
            ;;

        php)
            if [[ $prev == $cmd ]] ; then
				COMPREPLY=( $( compgen -W 'lamp' -- "$cur" ) )
				return 0
			fi
            case $subcmd in
                lamp)
					COMPREPLY=( $( compgen -W '$( _devassistant_baseopts_php )' -- "$cur" ) )
					return 0
					;;
            esac
            return 0
            ;;
    esac

    local split=false
    declare -F _split_longopt &>/dev/null && _split_longopt && split=true

    #_devassistant_complete_baseopts "$cur" "$prev" && return 0

    $split && return 0

    #if [[ $cur == -* ]] ; then
    #    COMPREPLY=( $( compgen -W '$( _devassistant_baseopts )' -- "$cur" ) )
    #    return 0
    #fi
    COMPREPLY=( $( compgen -W '${cmds[@]}' -- "$cur" ) )
} &&
complete -F _devassistant devassistant

# Local variables:
# mode: shell-script
# sh-basic-offset: 4
# sh-indent-comment: t
# indent-tabs-mode: nil
# End:
# ex: ts=4 sw=4 et filetype=sh
